scene/resources/surface_tool.cpp

```diff
@@ -135,18 +135,36 @@ bool SurfaceTool::Vertex::operator==(const Vertex &p_vertex) const {
 }
 
 uint32_t SurfaceTool::VertexHasher::hash(const Vertex &p_vtx) {
-	uint32_t h = hash_djb2_buffer((const uint8_t *)&p_vtx.vertex, sizeof(real_t) * 3);
-	h = hash_djb2_buffer((const uint8_t *)&p_vtx.normal, sizeof(real_t) * 3, h);
-	h = hash_djb2_buffer((const uint8_t *)&p_vtx.binormal, sizeof(real_t) * 3, h);
-	h = hash_djb2_buffer((const uint8_t *)&p_vtx.tangent, sizeof(real_t) * 3, h);
-	h = hash_djb2_buffer((const uint8_t *)&p_vtx.uv, sizeof(real_t) * 2, h);
-	h = hash_djb2_buffer((const uint8_t *)&p_vtx.uv2, sizeof(real_t) * 2, h);
-	h = hash_djb2_buffer((const uint8_t *)&p_vtx.color, sizeof(real_t) * 4, h);
-	h = hash_djb2_buffer((const uint8_t *)p_vtx.bones.ptr(), p_vtx.bones.size() * sizeof(int), h);
-	h = hash_djb2_buffer((const uint8_t *)p_vtx.weights.ptr(), p_vtx.weights.size() * sizeof(float), h);
-	h = hash_djb2_buffer((const uint8_t *)&p_vtx.custom[0], sizeof(Color) * RS::ARRAY_CUSTOM_COUNT, h);
-	h = hash_murmur3_one_32(p_vtx.smooth_group, h);
-	h = hash_fmix32(h);
+	uint32_t h = hash_murmur3_one_32(p_vtx.smooth_group);
+
+	h = hash_murmur3_one_float(p_vtx.color.r, h);
+	h = hash_murmur3_one_float(p_vtx.color.g, h);
+	h = hash_murmur3_one_float(p_vtx.color.b, h);
+	h = hash_murmur3_one_float(p_vtx.color.a, h);
+
+	h = hash_murmur3_one_real(p_vtx.normal.x, h);
+	h = hash_murmur3_one_real(p_vtx.normal.y, h);
+	h = hash_murmur3_one_real(p_vtx.normal.z, h);
+	h = hash_murmur3_one_real(p_vtx.binormal.x, h);
+	h = hash_murmur3_one_real(p_vtx.binormal.y, h);
+	h = hash_murmur3_one_real(p_vtx.binormal.z, h);
+	h = hash_murmur3_one_real(p_vtx.tangent.x, h);
+	h = hash_murmur3_one_real(p_vtx.tangent.y, h);
+	h = hash_murmur3_one_real(p_vtx.tangent.z, h);
+
+	h = hash_murmur3_one_real(p_vtx.uv.x, h);
+	h = hash_murmur3_one_real(p_vtx.uv.y, h);
+	h = hash_murmur3_one_real(p_vtx.uv2.x, h);
+	h = hash_murmur3_one_real(p_vtx.uv2.y, h);
+
+	h = hash_murmur3_one_real(p_vtx.vertex.x, h);
+	h = hash_murmur3_one_real(p_vtx.vertex.y, h);
+	h = hash_murmur3_one_real(p_vtx.vertex.z, h);
```

Comment:
This especially is really error prone

We might even want to add support for providing an existing hash value to the hash map methods to make this easier

Reply:
Do you mean hash caching?

Reply:
No I mean adding a seed to the methods to make chaining easier

Reply:
Perhaps something like:
```cpp
static _FORCE_INLINE_ uint32_t hash(const Vector3 &p_vec, uint32_t p_seed) {
	uint32_t h = hash_murmur3_one_real(p_vec.x, p_seed);
	h = hash_murmur3_one_real(p_vec.y, h);
	h = hash_murmur3_one_real(p_vec.z, h);
	return h;
}
static _FORCE_INLINE_ uint32_t hash(const Vector3 &p_vec) {
	return hash_fmix32(hash(p_vec, HASH_MURMUR3_SEED));
}
```
To avoid the mix step until the end